wgpu-core implementation for external textures

jamienicol opened the pull request:
**Connections**
Depends on #7822 
Part of #4386

**Description**
Add wgpu and wgpu-core implementation for external texture. wgpu-hal and naga implementations to follow, so for now this isn't particularly useful. But it does work with the noop backend, as can be seen in the added validation tests

**Testing**
Noop backend validation tests. Local testing. CTS testing can follow when more pieces are in place.

**Squash or Rebase?**

Rebase. If there are `fixup!` commits after review then squash them into the appropriate commit first

**Checklist**

- [x] Run `cargo fmt`.
- [x] Run `taplo format`.
- [x] Run `cargo clippy --tests`. If applicable, add:
  - [x] `--target wasm32-unknown-unknown`
- [x] Run `cargo xtask test` to run tests.
- [ ] If this contains user-facing changes, add a `CHANGELOG.md` entry. <!-- See instructions at the top of `CHANGELOG.md`. -->

Don't think this is changelog worthy yet

jamienicol committed:
[wgpu-core] Create default ExternalTextureParams buffer

In upcoming patches, wgpu will allowing the creation of bind groups
with either `TextureView`s or `ExternalTexture`s bound to a
`BindingType::ExternalTexture` bind group layout entry.

Wgpu-hal and the Naga-generated shaders must be able to handle both of
these cases. For external textures they will be provided a uniform
buffer containing the external texture's `ExternalTextureParams`. For
the texture view case, we must therefore provide the same.

To do this, we create a single buffer per device which can be shared
between all texture views. We initialize it with the required values
in Device::late_init_resources_with_queue(). We know that texture
views must have a single RGBA plane, with no rotation or
crop-rect. The only thing that can vary between them is their size. We
will therefore use the value of [0, 0] in the params buffer to
indicate to the shader that it should query the actual texture's size
rather than using the value provided in the buffer.

jamienicol committed:
[wgpu-core] Allow creation of bind groups containing external textures

Adds a `BindingResource` variant for external textures. In core's
create_bind_group() implementation, allow binding either external
textures or texture views to `BindingType::ExternalTexture` layout
entries.

In either case, provide HAL with a `hal::ExternalTextureBinding`,
consisting of 3 `hal::TextureBinding`s and a `hal::BufferBinding`. In
the texture view case we use the device's default params buffer for
the buffer. When there are fewer than 3 planes we can simply repeat an
existing plane multiple times - the contents of the params buffer will
ensure the shader only accesses the correct number of planes anyway.

Track the view or external texture in `BindGroupStates` to ensure they
remain alive whilst required.

And finally, add the corresponding API to wgpu, with an implementation
for the wgpu-core backend.

jamienicol committed:
[wgpu-core] Add ExternalTexture type to wgpu-core

`ExternalTexture` will form the basis of wgpu's implementation of
WebGPU's `GPUExternalTexture`. [1]

The application will be responsible for creating `Texture`(s) and
`TextureView`(s) from the external texture source and managing their
lifecycle. It may have a single RGBA texture, or it may have multiple
textures for separate Y and Cb/Cr planes. It can then create an external
texture by calling `create_external_texture()`, providing the texture
views and a descriptor. The descriptor provides the following required
information:

  * Whether the texture data is RGBA, or multiplanar or interleaved
    YCbCr.
  * The purpoted size of the external texture, which may not match the
    actual size of the underlying textures.
  * A matrix for converting from YCbCr to RGBA, if required.
  * A transform to apply to texture sample coordinates, allowing for
    rotation and crop rects.

The external texture stores a reference to the provided texture views,
and additionally owns a `Buffer`. This buffer holds data of the type
`ExternalTextureParams`, and will be provided as a uniform buffer to
shaders containing external textures. This contains information that
will be required by the shaders to handle external textures correctly.

Note that attempting to create an external texture will fail unless the
`Feature::EXTERNAL_TEXTURE` feature is enabled, which as of yet is not
supported by any HAL backends.

Additionally add the relevant API to wgpu, implemented for the
wgpu-core backend. The web and custom backends are unimplemented.

[1] https://www.w3.org/TR/webgpu/#gpuexternaltexture

jamienicol committed:
[wgpu] Add external texture validation tests

Adds validation tests using the noop backend covering creation of
external textures, and creation of bind groups containing external
textures.

teoxoy commented on wgpu-core/src/device/resource.rs:
```diff
@@ -173,6 +173,10 @@ pub struct Device {
+    /// Uniform buffer containing [`ExternalTextureParams`] with values such
+    /// that a [`TextureView`] bound to a [`wgt::BindingType::ExternalTexture`]
+    /// binding point will be rendered correctly.
+    pub(crate) default_external_texture_params_buffer: std::sync::OnceLock<Arc<Buffer>>,
```
Unfortunately the device can't hold wgpu-core buffers since buffers hold a strong ref to devices. This should be a hal buffer, you can use `zero_buffer` as inspiration.

jamienicol replied:
Makes sense, good catch. Fixed in e2375e4845d44c58ad162c197f1e1fa8fef3759c

teoxoy replied:
Fixup commit LGTM!

teoxoy reviewed (changes requested):
Looks good overall!

teoxoy commented:
This PR should be good to go but we should wait for https://github.com/gfx-rs/wgpu/pull/7822 to land first.

jamienicol commented:
In addition to waiting for https://github.com/gfx-rs/wgpu/pull/7822 to land first, let me rebase this manually too before merging. (I'm pretty sure my fixup commit needs split and squashed into two different ones in the series)

jamienicol commented:
I've rebased this, and squashed the fixup commit properly.

When adding tests for a PR which will follow, I noticed a slight issue. The `sample_transform` member of `ExternalTextureParams` is used to transform sample coordinates, allowing for things like crop rects, rotation, etc. I originally planned to use this for both samples and loads (doing some extra maths in the shader for loads). But my tests found a corner case which the CTS didn't, and I now realise we need a separate transform for loads. Less maths in the shader should also be better performance wise too. So I've updated `ExternalTextureParams` and `ExternalTextureDescriptor` to include that.

I'll also update #7826 accordingly

jamienicol commented:
Sorry for the noise, did not mean to link those issues!

Rebased patches, needed fixes to implement for WebGPUErrorType for CreateExternalTextureError, and for some changes with Hal::BufferBinding API.

also added functions for destroying external textures, and creating external textures in error state

jimblandy commented:
Since I've landed #7822, I rebased this on trunk.
